File: 0501~1000/0823. Binary Trees With Factors/main.py
"""
Source: https://rb.gy/jmhx8
Date: 2023/8/15
Skill:
Ref:
Runtime: 287 ms, faster than 86.14%
Memory Usage: 17.72 MB, less than 14.46%
Time complexity:
Space complexity:
Constraints:
    1 <= arr.length <= 1000
"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        res, mod, num_set, sz = 0, 10 ** 9 + 7, set(arr), len(arr)
        product, num2idx = [[] for _ in range(sz)], defaultdict(int)
        arr.sort()
        for i in range(sz):
            num2idx[arr[i]] = i
        for i in range(sz - 1, -1, -1):
            for j in range(i - 1, -1, -1):
                if arr[i] % arr[j] == 0 and arr[i] // arr[j] in num_set:
                    if arr[i] // arr[j] > arr[j]:
                        break
                    product[i].append([num2idx[arr[i] // arr[j]], j])

        dp = [-1 for _ in range(sz)]

        def helper(idx):
            if dp[idx] != -1:
                return dp[idx]
            cnt = 1
            for i in range(len(product[idx])):
                if product[idx][i][0] != product[idx][i][1]:
                    cnt += 2 * helper(product[idx][i][0]) * helper(product[idx][i][1])
                else:
                    cnt += helper(product[idx][i][0]) * helper(product[idx][i][1])
            dp[idx] = cnt
            return cnt

        for i in range(sz):
            res += helper(i)
        return res % mod


if __name__ == "__main__":
    s = Solution()

File: 0501~1000/0823. Binary Trees With Factors/main.cpp
/**
 * Source: tinyurl.com/yrhwrfud
 * Date: 2023/10/26
 * Skill:
 * Runtime: 47 ms, faster than 63.39% of C++ online submissions
 * Memory Usage: 9.39 MB, less than 70.87% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>

using namespace std;

#define ll long

class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr) {
        ll res = 0, mod = 1e9+7;
        sort(arr.begin(), arr.end());
        unordered_map<int, ll> cnt;
        for (int i = 0; i < arr.size(); i++) {
            ll num = 1;
            int limit = sqrt(arr[i]);
            for (int j = i - 1; j >= 0; j--) {
                if (arr[j] <= limit) {
                    if (arr[j] * arr[j] == arr[i]) {
                        num = (num + cnt[arr[j]] * cnt[arr[j]] % mod) % mod;
                    }
                    break;
                }else if (arr[i] % arr[j] == 0 && cnt.find(arr[i] / arr[j]) != cnt.end()) {
                    num = (num + cnt[arr[j]] * cnt[arr[i] / arr[j]] * 2 % mod) % mod;
                }
            }
            res = (res + num) % mod;
            cnt[arr[i]] = num;
        }
        return res;
    }
};


int main() {
    Solution s;
    vector<int> nums{-1,0,1,0,3,3};
    vector<vector<int>> arr{{1,5}, {10,11},{12,18},{20,25},{30,32}};
}